### ISSUE TITLE
Default alignment for DynamicStorageBuffer is incorrect.

### ISSUE THREAD

**Issue by AndrewB330** [bevyengine/bevy#5411]
I'm using a `DynamicStorageBuffer` to store a lot of `u32`s and use them in compute shader. I noticed that in `DynamicStorageBuffer::default()` we are initializing `scratch` with `DynamicStorageBufferWrapper` with default alignment that is set to 256 by default (in encase library):
```rs
pub const fn new(buffer: B) -> Self {
    Self::new_with_alignment(buffer, 256) // default alignment in wrapper
}
```
I think this is not correct, since we can easily have types with smaller sizes, after I changed `scratch` initialization to `DynamicStorageBufferWrapper::new_with_alignment(Vec::new(), 4)` it seems that everything working fine now. I would create a PR but I'm not sure what is the correct number here, since it does not represent shader type alignment but something else, because it should be a power of 2.

**Comment by superdump** [comment]
It isn't incorrect, it is just unnecessarily conservative.

The alignment required for dynamic offsets in uniform / storage buffers depends on the underlying graphics API / GPU drivers. The actual requirement is in the `RenderDevice::limits() -> wgpu::Limits`. For uniform buffers it is the `min_uniform_buffer_offset_alignment` field and for storage buffers it is the `min_storage_buffer_offset_alignment` field. Ideally we should default to use whatever value is expressed in these members of `wgpu::Limits`.

**Comment by AndrewB330** [comment]
@superdump 
Hm, but it doesn't work with my shader, so I assume it is incorrect. If it was correct but conservative then it would work always but with some perf/memory implications, right? I'm compiling glsl shader with one buffer:
```glsl
layout (std430, binding = 1) buffer Materials {
    uint materials[];
};
```
so that elements have alignment of 4 bytes, and not 256. So after I bind a buffer here I'm getting an element only at i * 64 positions.

**Comment by superdump** [comment]
The 256 isn't about the internal buffer alignment. That is defined by std430/wgsl and is handled by the `encase` crate via the bevy `StorageBuffer`/`DynamicStorageBuffer` abstractions. The 256 is about the alignment of a dynamic offset that is passed when binding a storage buffer that is configured to use dynamic offsets.

Could you share your code so I can take a look? What are you using as the type within the `DynamicStorageBuffer`?

**Comment by AndrewB330** [comment]
@superdump yes, I am in fact using it like `DynamicStorageBuffer<u32>` because this was my understanding of it, since I've seen similar APIs in other libraries (e.g. vulkano). It also seems that I can't derive `ShaderType` on a struct with `Vec` inside. I think `derive(ShaderType)` can be use only on types with known size, from my understanding.

Here is my code:

```rs
#[derive(Default)]
pub struct Buffers {
    pub materials: DynamicStorageBuffer<u32>,
}

// Layout definition
BindGroupLayoutEntry {
    binding: 11,
    visibility: ShaderStages::COMPUTE,
    ty: BindingType::Buffer {
        ty: BufferBindingType::Storage { read_only: false },
        has_dynamic_offset: false,
        min_binding_size: None,
    },
    count: None,
},

// Binding creation:
BindGroupEntry {
    binding: 11,
    resource: buffers.materials.binding().unwrap(),
},
```

```glsl
// Shader
layout (std430, binding = 11) buffer Materials {
    uint materials[];
};
```

**Comment by superdump** [comment]
That isn't how `DynamicStorageBuffer` works. It does not allow you to make a single buffer binding and index into an array in the shader.

It looks like `encase` supports runtime-sized arrays with an annotation and allows you to get an array length in a member like this:
```rust
#[derive(ShaderType)]
struct Materials {
    length: ArrayLength,
    #[size(runtime)]
    materials: Vec<u32>,
}

struct Something {
    buffer: StorageBuffer<Materials>,
}
```

And whether you use `StorageBuffer` or `DynamicStorageBuffer` is about whether you need different arrays of u32 for different draw commands. If you use one array of u32 for all your draw commands, then you only need `StorageBuffer`.

**Comment by cshenton-work** [comment]
Nothing in the documentation for `DynamicStorageBuffer` suggests it works the way it does. I also have, based on its description (and the fact that I don't create it with an explicit alignment like `DynamicUniformBuffer`), proceeded with the assumption it is simply a local append buffer with natural type alignment that I can dynamically append to and then upload.

So yeah, decent amount of time wasted due to this undocumented and unintuitive behaviour. Please document it / change it to behave how most people would expect it to.
